metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.server.common.MetadataVersion;
+import org.apache.kafka.common.utils.Time;
+
+import java.nio.charset.StandardCharsets;
+import javax.crypto.spec.SecretKeySpec;
+import javax.crypto.Mac;
+
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+
+import static org.apache.kafka.common.protocol.Errors.DELEGATION_TOKEN_AUTH_DISABLED;
+import static org.apache.kafka.common.protocol.Errors.DELEGATION_TOKEN_EXPIRED;
+import static org.apache.kafka.common.protocol.Errors.DELEGATION_TOKEN_NOT_FOUND;
+import static org.apache.kafka.common.protocol.Errors.DELEGATION_TOKEN_OWNER_MISMATCH;
+import static org.apache.kafka.common.protocol.Errors.INVALID_PRINCIPAL_TYPE;
+import static org.apache.kafka.common.protocol.Errors.NONE;
+import static org.apache.kafka.common.protocol.Errors.UNSUPPORTED_VERSION;
+
+/**
+ * Manages DelegationTokens.
+ */
+public class DelegationTokenControlManager {
+    private Time time = Time.SYSTEM;
+
+    static class Builder {
+        private LogContext logContext = null;
+        private DelegationTokenCache tokenCache = null;
+        private String secretKeyString = null;
+        private long tokenDefaultMaxLifetime = 0;
+        private long tokenDefaultRenewLifetime = 0;
+
+        Builder setLogContext(LogContext logContext) {
+            this.logContext = logContext;
+            return this;
+        }
+
+        Builder setTokenCache(DelegationTokenCache tokenCache) {
+            this.tokenCache = tokenCache;
+            return this;
+        }
+
+        Builder setTokenKeyString(String secretKeyString) {
+            this.secretKeyString = secretKeyString;
+            return this;
+        }
+
+        Builder setDelegationTokenMaxLifeMs(long tokenDefaultMaxLifetime) {
+            this.tokenDefaultMaxLifetime = tokenDefaultMaxLifetime;
+            return this;
+        }
+
+        Builder setDelegationTokenExpiryTimeMs(long tokenDefaultRenewLifetime) {
+            this.tokenDefaultRenewLifetime = tokenDefaultRenewLifetime;
+            return this;
+        }
+
+        DelegationTokenControlManager build() {
+            if (logContext == null) logContext = new LogContext();
+            return new DelegationTokenControlManager(
+              logContext,
+              tokenCache,
+              secretKeyString,
+              tokenDefaultMaxLifetime,
+              tokenDefaultRenewLifetime);
+        }
+    }
+
+    private final Logger log;
+    private final DelegationTokenCache tokenCache;
+    private final String secretKeyString;
+    private final long tokenDefaultMaxLifetime;
+    private final long tokenDefaultRenewLifetime;
+
+    private DelegationTokenControlManager(
+        LogContext logContext,
+        DelegationTokenCache tokenCache,
+        String secretKeyString,
+        long tokenDefaultMaxLifetime,
+        long tokenDefaultRenewLifetime
```

Comment:
Might be good to append `Ms` to clearly indicate the units.  Same elsewhere if necessary.

Reply:
Yeah, I added it everywhere.